File: submit_20176342_hw2/EasyUDPClient.java
package submit_20176342_hw2;

import java.util.*;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.Socket;
import java.net.SocketException;
import java.net.UnknownHostException;

public class EasyUDPClient {

    final static String serverName = "nsl2.cau.ac.kr";
    final static int serverPort = 26342;

    public static class ByeByeThread extends Thread{
        DatagramSocket conn;

        ByeByeThread(DatagramSocket conn){
            this.conn = conn;
        }

        public void run() {
            try {
                Thread.sleep(200);
                byebye();
                this.conn.close();
               
            
                //some cleaning up code...

            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                e.printStackTrace();
            }
        }
    }

    public static void main(String[] args){

        
        DatagramSocket conn = null;

        try{
            conn = new DatagramSocket();
            Runtime.getRuntime().addShutdownHook(new ByeByeThread(conn));


            // OutputStream os = conn.getOutputStream();
            // InputStream is = conn.getInputStream();

            HandleInput H = new HandleInput(conn);
            H.start();

            // conn.close();

        } catch (SocketException e) {
             e.printStackTrace(); 
        } catch (SecurityException e) {
            e.printStackTrace(); 
        }



        


    }

    
    public static void byebye(){
        System.out.println("Bye bye~");
    }

    public static void processOption(int opt, DatagramSocket conn){
        long startTime = new Date().getTime();

        String requestString;
        String reply;

        switch(opt){
            case 1:
                System.out.printf("Input lowercase sentence: ");
                Scanner sc = new Scanner(System.in);
                String input = sc.nextLine();
                requestString = opt+"|"+input;

                startTime = new Date().getTime();

                sendPacket(conn, requestString);
        
                reply = readPacket(conn);

                System.out.printf("Reply from server: %s\n",reply);
                break;
            case 2:
                requestString = opt+"|";
                sendPacket(conn, requestString);
                reply = readPacket(conn);
                System.out.printf("Reply from server: client IP = %s, port = %s\n",reply.split(":")[0], reply.split(":")[1]);
                break;
            case 3:
                requestString = opt+"|";
                sendPacket(conn, requestString);
                reply = readPacket(conn);
                System.out.printf("Reply from server: requests served = %s\n",reply);
                break;
            case 4:
                requestString = opt+"|";
                sendPacket(conn, requestString);
                reply = readPacket(conn);
                printDuration(reply);
                break;
            case 5:
                conn.close();
                System.exit(0);
                break;
            default:
                conn.close();
                System.exit(0);
                break;
   
   
        }

    

        
        
        printRTT(startTime);
    }


    public static void printRTT(long startTime){
        System.out.printf("RTT = %dms\n\n\n", (new Date().getTime()-startTime));
    }

    public static void printDuration(String t){
        int h = 0;
        int m = 0;
        int s = 0;
        
        if(t.contains("h")){
            h = Integer.parseInt( t.split("h")[0] );
            t = t.split("h")[1];
        }

        if(t.contains("m")){
            m = Integer.parseInt( t.split("m")[0] );
            t = t.split("m")[1];
        }

        if(t.contains("s")){
            s = Integer.parseInt( t.split("s")[0] );
        }

        System.out.printf("Reply from server: run time = %02d:%02d:%02d\n", h, m, s);
    }

    public static void sendPacket(DatagramSocket conn, String requestString){
        try{
            InetAddress serverIP = InetAddress.getByName(serverName);

            DatagramPacket dp = new DatagramPacket(requestString.getBytes(),requestString.getBytes().length,serverIP, serverPort);
             
            //epdlxj wjsthd
            conn.send(dp);
            
        } catch(IOException e){
            //do nothing
        }
        
    }

    public static String readPacket(DatagramSocket conn){
        try
        {
            byte[] buffer = new byte[1024];
            DatagramPacket response = new DatagramPacket(buffer, buffer.length);
            conn.receive(response);
            String res = new String(buffer, 0, response.getLength());

            return res;
        
        } catch(IOException e){
        
            return null;
        }

        

    }

    public static void printOption(){
        System.out.printf("<Menu>\n");
        System.out.printf("option 1) convert text to UPPER-case letters.\n");
        System.out.printf("option 2) ask the server what the IP address and port number of the client is.\n");
        System.out.printf("option 3) ask the server how many client requests(commands) it has served so far.\n");
        System.out.printf("option 4) ask the server program how long it has been running for since it started.\n");
        System.out.printf("option 5) exit client program\n");
    }


    static class HandleInput extends Thread {
        DatagramSocket conn;
        
        HandleInput(DatagramSocket conn){
            this.conn = conn;
        }
        public void run(){
            while(true){
                try{
                    printOption();
                    System.out.printf("Please select your option :");
                    Scanner sc = new Scanner(System.in);
                    int opt = sc.nextInt();
                    processOption(opt, this.conn);
                } catch(NoSuchElementException e){
                    System.exit(0);
                } catch(NullPointerException e){
                    System.exit(0);
                }
                
            }
            
        }
    }

    static class HandleReceive extends Thread {

    }

}

File: submit_20176342_hw2/EasyUDPClient.go
/**
 * 20176342 Song Min Joon
 * EasyUDPClient.go
 **/

package main

import (
	"bufio"
	"fmt"
	"net"
	"os"
	"os/signal"
	"strconv"
	"strings"
	"syscall"
	"time"
)

var serverName string = "nsl2.cau.ac.kr" //server host
var serverPort string = "26342"          //server port

func main() {

	c := make(chan os.Signal, 2)
	signal.Notify(c, os.Interrupt, syscall.SIGTERM) // for exit the program gracefully
	go func() {
		for sig := range c {
			_ = sig
			byebye() // print byebye func
		}
	}()

	conn, err := net.ListenPacket("udp", ":")

	if err != nil {
		fmt.Printf("Please check your server is running\n")
		return
	}
	defer conn.Close()

	localAddr := conn.LocalAddr().(*net.UDPAddr) // get local addr

	fmt.Printf("Client is running on port %d\n", localAddr.Port)

	server_addr, _ := net.ResolveUDPAddr("udp", serverName+":"+serverPort) // make path for give and take packet with server

	for {
		//infinite loop for input option
		handleInput(conn, server_addr)
	}

}

func byebye() {
	//print Bye bye
	fmt.Printf("Bye bye~\n")
	os.Exit(0)
}

func handleInput(conn net.PacketConn, addr *net.UDPAddr) {
	/*
		this function prints Menu and 5 Options and wait for input Option.
	*/
	printOption()
	fmt.Printf("Please select your option :")
	var opt int
	fmt.Scanf("%d", &opt)
	processOption(opt, conn, addr)
}

func handleError(conn net.PacketConn, errmsg string) {
	if conn != nil {
		conn.Close()
	}
	fmt.Println(errmsg)
}

func processOption(opt int, conn net.PacketConn, addr *net.UDPAddr) {
	/*
		if option is given, it sends packet to server and get response.
	*/
	startTime := time.Now() //startTime for print RTT

	// var temp int
	// fmt.Scanf("%s", &temp)

	switch opt {
	case 1:
		// Option 1

		fmt.Printf("Input lowercase sentence: ")
		input, _ := bufio.NewReader(os.Stdin).ReadString('\n')
		startTime = time.Now()
		requestString := strconv.Itoa(opt) + "|" + input
		sendPacket(conn, requestString, addr)
		buffer := make([]byte, 1024)
		bufferSize := readPacket(conn, &buffer)
		fmt.Printf("Reply from server: %s\n", string(buffer[:bufferSize]))

	case 2:
		// Option 2

		requestString := strconv.Itoa(opt) + "|"
		sendPacket(conn, requestString, addr)
		buffer := make([]byte, 1024)
		bufferSize := readPacket(conn, &buffer)
		fmt.Printf("Reply from server: client IP = %s, port = %s\n", string(strings.Split(string(buffer[:bufferSize]), ":")[0]), string(strings.Split(string(buffer[:bufferSize]), ":")[1]))

	case 3:
		// Option 3

		requestString := strconv.Itoa(opt) + "|"
		sendPacket(conn, requestString, addr)
		buffer := make([]byte, 1024)
		bufferSize := readPacket(conn, &buffer)
		fmt.Printf("Reply from server: requests served = %s \n", string(buffer[:bufferSize]))

	case 4:
		// Option 4

		requestString := strconv.Itoa(opt) + "|"
		sendPacket(conn, requestString, addr)
		buffer := make([]byte, 1024)
		bufferSize := readPacket(conn, &buffer)
		timeD, _ := time.ParseDuration(string(buffer[:bufferSize]))

		printDuration(timeD)

	case 5:
		// Option 5

		byebye()
		conn.Close()
		os.Exit(0)
	default:
		// not Option 1~5 (default)

		byebye()
		conn.Close()
		os.Exit(0)
	}
	printRTT(time.Since(startTime))// print RTT Since startTime

}

func printRTT(d time.Duration) {
	//print RTT Time between before send packet and after send packet in Milliseconds.

	d = d.Round(time.Millisecond)

	s := d / time.Millisecond

	fmt.Printf("RTT = %dms \n\n\n", s)

}
func printDuration(d time.Duration) {
	//print server running time in proper form(HH:MM:ss)

	d = d.Round(time.Second)

	h := d / time.Hour
	d -= h * time.Hour

	m := d / time.Minute
	d -= m * time.Minute

	s := d / time.Second
	d -= s * time.Second

	fmt.Printf("Reply from server: run time = %02d:%02d:%02d\n", h, m, s)
}

func sendPacket(conn net.PacketConn, requestString string, addr *net.UDPAddr) {
	//send Packet to server

	conn.WriteTo([]byte(requestString), addr)
}

func readPacket(conn net.PacketConn, buffer *[]byte) int {
	//read Packet from server and saves to buffer and return buffer size.

	//There is no way to connection is established succesfully or connection is disconnect
	//cause there is no connection in UDP
	count, _, _ := conn.ReadFrom(*buffer)
	return count
}

func printOption() {
	//print Menu and 5 Options.

	fmt.Printf("<Menu>\n")
	fmt.Printf("option 1) convert text to UPPER-case letters.\n")
	fmt.Printf("option 2) ask the server what the IP address and port number of the client is.\n")
	fmt.Printf("option 3) ask the server how many client requests(commands) it has served so far.\n")
	fmt.Printf("option 4) ask the server program how long it has been running for since it started.\n")
	fmt.Printf("option 5) exit client program\n")

}
